### PR TITLE
KAFKA-6046 DeleteRecordsRequest to a non-leader should give proper error

### PR DESCRIPTION
**Pull request by tedyu**

### CODE REVIEW

**Comment by tedyu** [comment]
AgentTest.testAgentActivatesFaults failure was not related

**Comment by tedyu** [comment]
@tombentley 
Can you take a look ?

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 1]
```diff
@@ -515,6 +515,11 @@ class ReplicaManager(val config: KafkaConfig,
             case Some(p) =>
               if (p eq ReplicaManager.OfflinePartition)
                 throw new KafkaStorageException("Partition %s is in an offline log directory on broker %d".format(topicPartition, localBrokerId))
+              p.leaderReplicaIfLocal match {
+                case Some(_) =>
```
Comment (tombentley) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: You should be able to use `case Some(_) => p` and thus omit the `p` that's currently the last expression in the block.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 2]
```diff
@@ -515,6 +515,11 @@ class ReplicaManager(val config: KafkaConfig,
...
+                case None =>
+                  throw new NotLeaderForPartitionException("Broker %d is not leader for partition %s".format(localBrokerId, topicPartition))
```
Comment (tombentley) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: I understand that using Scala template strings (`s"Broker $localBrikerId is not leader..."`) is preferred over `String.format()`

Comment (tombentley) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: Also, long line, but using templates string might fix that.

**Comment by tombentley** [comment]
@tedyu there will need to be a test, but there don't seem to be any tests for `DeleteRecordsRequest` yet. KIP-205 will add API to `AdminClient` (thus it could be tested from `AdminClientIntegrationTest`). For the work on KIP-201 I'm planning to add a lower level test too, so if you're amenable, I could rebase my work onto this patch and add the test?

**Comment by tombentley** [comment]
Sorry, I meant KIP-204, ftr.

**Comment by tedyu** [comment]
If you add the test in your PR, that would be nice.
Thanks

**Comment by tedyu** [comment]
@tombentley
Do you have other comment ?

**Comment by tombentley** [comment]
lgtm.

I will try to open a PR with a test ASAP, but since KIP-201 is still being voted on it will be next week at the earliest, assuming no one asks for revisions during voting, and assuming the vote passes.

**Comment by tedyu** [comment]
@onurkaraman @ijuma @guozhangwang 
Can you take a look ?

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 3]
```diff
@@ -515,7 +515,11 @@ class ReplicaManager(val config: KafkaConfig,
             case Some(p) =>
               if (p eq ReplicaManager.OfflinePartition)
                 throw new KafkaStorageException("Partition %s is in an offline log directory on broker %d".format(topicPartition, localBrokerId))
-              p
+              p.leaderReplicaIfLocal match {
+                case Some(_) => p
+                case None =>
+                  throw new NotLeaderForPartitionException(s"Broker $localBrokerId is not leader for partition $topicPartition")
+              }
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: We can use `getLeaderReplicaIfLocal` to simplify the code.

**Comment by ijuma** [comment]
There are tests in `LegacyAdminClientTest`, we should be able to add a test there.

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 3 above)
Comment (tedyu) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: getLeaderReplicaIfLocal returns replica.
The return value here is partition.
I feel the current formation is clearer.
See the code for convertedOffset below which has the same structure.

**Comment by tedyu** [comment]
For LegacyAdminClientTest, looks like record deletion is directed to the correct leader.
Mind sharing some hint how to trigger the new code added ?

Thanks

**Comment by tombentley** [comment]
The tests in `LegacyAdminClientTest` won't work because the legacy `AdminClient` first makes a metadata request and returns `LEADER_NOT_AVAILABLE` for any partition without a leader, thus is not exercising the broker-side code @tedyu has added.

The work I've done for KIP-201 adds a network level test like the `CreateTopicsRequestTest`, so I can test it there if you're prepared to wait, or with a bit of work I could pull just that test out and add it to this PR if you prefer @ijuma?

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 3 above)
Comment (ijuma) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: The problem is that the same logic exists in many places leading to bugs. The `convertedOffset` code only needs a replica. However, it's true that we need the `partition` for `deleteRecordsOnLeader`. My suggestion is to introduce a new method and use it in `getReplicaLeaderIfLocal`:

```scala
  def getLeaderReplicaIfLocal(topicPartition: TopicPartition): Replica =  {
    val (_, replica) = getPartitionAndLeaderReplicaIfLocal(topicPartition)
    replica
  }

  def getPartitionAndLeaderReplicaIfLocal(topicPartition: TopicPartition): (Partition, Replica) =  {
    val partitionOpt = getPartition(topicPartition)
    partitionOpt match {
      case None =>
        throw new UnknownTopicOrPartitionException(s"Partition $topicPartition doesn't exist on $localBrokerId")
      case Some(partition) =>
        if (partition eq ReplicaManager.OfflinePartition)
          throw new KafkaStorageException(s"Partition $topicPartition is in an offline log directory on broker $localBrokerId")
        else partition.leaderReplicaIfLocal match {
          case Some(leaderReplica) => (partition, leaderReplica)
          case None =>
            throw new NotLeaderForPartitionException(s"Leader not local for partition $topicPartition on broker $localBrokerId")
        }
    }
  }
```

**Comment by ijuma** [comment]
@tombentley once this PR is ready, you can just verify that your failing test passes with these changes and we can then integrate the test as part of your PR. We should add a note about this in the PR description.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 4]
```diff
@@ -515,7 +520,8 @@ class ReplicaManager(val config: KafkaConfig,
             case Some(p) =>
               if (p eq ReplicaManager.OfflinePartition)
                 throw new KafkaStorageException("Partition %s is in an offline log directory on broker %d".format(topicPartition, localBrokerId))
-              p
+              val (partition, _) = getPartitionAndLeaderReplicaIfLocal(topicPartition)
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: You can remove a bunch of the code in this method now, `getPartition`, `partition.leaderReplicaIfLocal`, the `OfflinePartition` check. All of that is done in `getPartitionAndLeaderReplicaIfLocal`.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 5]
```diff
@@ -511,23 +516,10 @@ class ReplicaManager(val config: KafkaConfig,
...
           val convertedOffset =
             if (requestedOffset == DeleteRecordsRequest.HIGH_WATERMARK) {
-              partition.leaderReplicaIfLocal match {
-                case Some(leaderReplica) =>
-                  leaderReplica.highWatermark.messageOffset
-                case None =>
-                  throw new NotLeaderForPartitionException("Leader not local for partition %s on broker %d"
-                    .format(topicPartition, localBrokerId))
-              }
+              replica.highWatermark.messageOffset
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:522]: Seems like we can remove the comment:

" // NOTE: Failed produce requests metric is not incremented for known exceptions
  // it is supposed to indicate un-expected failures of a broker in handling a produce request"

**Review by ijuma — COMMENTED** [review]
Thanks for the updates. One additional suggestion for existing code. The rest looks good.

**Review by ijuma — APPROVED** [review]
LGTM if @tombentley can confirm that the his protocol test passes with this change.

**Comment by tedyu** [comment]
ResetIntegrationTest failure is not related.

**Comment by tombentley** [comment]
@tedyu I rebased my WIP for KIP-201 onto your commits, but the test still failed because the call to `getPartition(topicPartition)` in `getPartitionAndLeaderReplicaIfLocal()` throws `UnknownTopicOrPartitionException` before we can enter the `match` expression.

I've pushed my original WIP [here](https://github.com/tombentley/kafka/tree/KAFKA-5693-rationalise-policy-interfaces) (that branch lacks your changes) and [here](https://github.com/tombentley/kafka/tree/KAFKA-5693-rationalise-policy-interfaces) (the rebase of my work on top of yours).

You can try to use my `DeleteRecordsRequestTest.testNotLeader()` if you want. The test currently expects `UNKNOWN_TOPIC_OR_PARTITION`, with your change it should expect `NOT_LEADER_FOR_PARTITION`.

**Comment by tedyu** [comment]
```
    val request = new DeleteRecordsRequest.Builder(1000,
      Map[TopicPartition, java.lang.Long](topicPartition -> 5L).asJava, false).build()
```
The above seems to be the same as Single topic in testValidDeleteRecordsRequests (apart from timeout value)
Mind letting me know the difference ?

Thanks

**Comment by tedyu** [comment]
In createTopic():
```
    // create topic
    AdminUtils.createOrUpdateTopicPartitionAssignmentPathInZK(zkUtils, topic, partitionReplicaAssignment)
```
Partition assignment is supposed to be written in ZK.
Isn't this supposed to let follower have information on the partition ?

**Comment by tedyu** [comment]
Looks like the following call to makeFollowers() in ReplicaManager#becomeLeaderOrFollower() isn't made in test:
```
        val partitionsBecomeFollower = if (partitionsToBeFollower.nonEmpty)
          makeFollowers(controllerId, controllerEpoch, partitionsToBeFollower, correlationId, responseMap)
```

**Comment by ijuma** [comment]
It would be good to verify if the test is allowing enough time for the metadata to propagate. If the broker doesn't know about the partition, then it is expected that it will return `UNKNOWN_TOPIC_OR_PARTITION`.

**Comment by tombentley** [comment]
I added a commit to both my branches adding a call to `TestUtils.waitUntilMetadataIsPropagated()` before making the `DeleteRecordsRequest` and the response is still `UNKNOWN_TOPIC_OR_PARTITION`.

**Comment by tedyu** [comment]
As I stated earlier, we need to trigger follower code path (ReplicaManagerTest.testClearPurgatoryOnBecomingFollower does).

I am attempting this:

https://pastebin.com/cC737EWh

However, it does't compile yet:

https://pastebin.com/t5nkUmc8

I want to get reviewers' opinion before I go further along this route.

**Comment by tedyu** [comment]
What if we use the following to trigger the deletion 
```
      rm.deleteRecords(1000,
        Map[TopicPartition, scala.Long](new TopicPartition(topic, 0) -> 5L), null)
```

**Review by ijuma — APPROVED** [review]
I looked into this and the issue is that the current behaviour for the produce request (which we are now also following for delete records) is not exactly what Tom was expecting. The behaviour is:

1. If the broker is a follower, return NOT_LEADER.
2. If the broker is the leader, don't return an error.
3. If the broker is neither a follower nor a leader, return UNKNOWN_TOPIC_OR_EXCEPTION. Tom's test expected NOT_LEADER to be returned for this case.

So, this PR LGTM in that it makes delete records consistent with produce.

**Comment by ijuma** [comment]
retest this please

**Comment by asfgit** [comment]
FAILURE
 7973 tests run, 5 skipped, 2 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 7973 tests run, 5 skipped, 1 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8081 tests run, 5 skipped, 0 failed.
--none--

**Pull request closed without merging**